### ISSUE TITLE
Golang client for NATS web socket

### ISSUE THREAD

**Issue by gedw99** [nats-io/nats.go#831]

**Comment by gedw99** [comment]
https://github.com/nats-io/nats.go/issues/530

Previous issue on this.

**Comment by kozlovic** [comment]
> i was told via the nats.ws repo that the nats.go client may support compiling to wasm .

No, the Go client has native Websocket support. That is, use `ws://` or `wss://` as the URL scheme and the Go library will send/read websocket frames natively (no need for a proxy, etc..). Support was added in v1.11.0, but some bugs have been fixed in later releases, so would recommend using latest, which at the time of this writing is v1.12.3.

**Comment by gedw99** [comment]
It would need to support web sockets and use go build tags so it could connect to the golang server

**Comment by kozlovic** [comment]
Sorry, I don't understand your last comment.

**Comment by gedw99** [comment]
Well basically it has WSS and WS support so if we started using go build tags then we could compile it for running as wasm or as a normal golang binary right ?

**Comment by kozlovic** [comment]
This repository is a library, so there is no binary. From what I understand, this cannot be compiled as wasm because there is currently no support for threads and network, which is used by the library.

**Comment by gedw99** [comment]
I know it’s a library / package . But I obviously meant imported and then compiled to wasm.

about the threads. there are no threads in golang. You mean concurrency management ?

regarding web socket networking it looks like the team rolled their own. Which would mean likely small modification to allow it to work

**Comment by gedw99** [comment]
It’s really disappointing  that the nats  team does not want to support wasm .  

It’s exploding . Have a look at gio that allows building web client in golang that compile to wasm and competes with flutter in terms of the architecture

**Comment by kozlovic** [comment]
> about the threads. there are no threads in golang. You mean concurrency management ?

There are, just not exposed. The library uses various go routines, and likely to run on different threads.

> regarding web socket networking it looks like the team rolled their own. Which would mean likely small modification to allow it to work

What modifications though? Again, I think that there is no network support in wasm.. but again, I am personally not familiar, and although wasm support is in the roadmap, we are not there yet: https://github.com/nats-io/nats-general

**Comment by kozlovic** [comment]
> It’s really disappointing that the nats team does not want to support wasm .

Never said that. Look at the link I just sent. In the roadmap, but we are a small team and have lot to do... so please bear with us.

**Comment by gedw99** [comment]
Ok

about network support .

you have web sockets available

gorilla lib does not work. 

Nooby web socketblib does work and is designed for wasm and non wasm .

here: https://pkg.go.dev/nhooyr.io/websocket

**Comment by gedw99** [comment]
Gio does the same go build tag approach as the web socket lib 

https://github.com/gioui/gio-x/tree/main/explorer

Ex: https://github.com/nhooyr/websocket/blob/master/compress_notjs.go

**Comment by ripienaar** [comment]
@gedw99 we have a vast WASM roadmap that affects many parts of our eco system including the server, it's something we are VERY keen on getting.  But even more keen on getting right, consistent and easy to support long term for us and the community.

So while yes, in isolation, it's easy to point at some random WASM repos and go why can't you just do that.  But the problem is actually quite vast as we aim to provide a server side and client side WASM story and we have to do so within our security model and more.

We are in fact very keen to do it, but for our full purposes, it's a bit easy in the life of WASM - but we have it on the roadmap and keep following the WASM space and refining our thoughts.

**Comment by gedw99** [comment]
I hear you and saw it’s slotted in for the 2nd quarter of 2022 on your roadmap.

I just wanted to let you know from a user that it would be highly useful. 
With  WASM technology maturing and all the opportunities it provides for use developers to innovate. 

Gio is a great example of how mature it is. Not perfect but definitely workable.

Just wanted to point out that library in case you were not aware of it. No dependencies. Mit license.

**Comment by gedw99** [comment]
I think in the mean time the only way forward for me is to write a proxy that sits in between the gio gui and the nats server.

**Comment by ripienaar** [comment]
depends on your needs exactly but https://resgate.io/ might work

**Comment by kozlovic** [comment]
> Just wanted to point out that web socket library in case you were not aware of it. No dependencies. Mit license.

I was personally not, and for sure no dependency (except for compression stuff, we we also use that one in the server), and MIT licenses are good indeed. Would have to see how "plug-able" it would be in our library, which may or may not be easy, and again, would take time that we are a bit lacking at the moment. But again, thanks for the pointer.

**Comment by gedw99** [comment]
I know resgate well. Does not quite fit the need I have. But thanks for pointing it out
